File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case02_transformation/Case09_Iterate.scala
package com.yw.flink.example.scalacases.case02_transformation

import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

/**
  *
  * @author yangwei
  */
object Case09_Iterate {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._
    val ds: DataStream[String] = env.socketTextStream("node5", 9999)
    val ds2: DataStream[Int] = ds.map(_.toInt)

    val result: DataStream[Int] = ds2.iterate(iteration => {
      //定义迭代体
      val minusOne: DataStream[Int] = iteration.map(v => {
        println("迭代体中进入的数据为 = " + v)
        v - 1
      })

      //定义迭代条件
      val cn: DataStream[Int] = minusOne.filter(_ > 0)

      //定义哪些数据最后输出
      val end: DataStream[Int] = minusOne.filter(_ != 0)

      (cn, end)
    })

    result.print()
    env.execute()

  }

}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case02_transformation/Case09_Iterate.java
package com.yw.flink.example.javacases.case02_transformation;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.IterativeStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * 用于迭代计算，迭代计算场景，注重的是某块业务逻辑一直迭代计算
 */
public class Case09_Iterate {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        DataStreamSource<String> ds = env.socketTextStream("node5", 9999);
        SingleOutputStreamOperator<Integer> ds2 = ds.map(one -> Integer.valueOf(one));

        //创建迭代流
        IterativeStream<Integer> iterate = ds2.iterate();

        //定义迭代体 ，就是我们看中的多次迭代计算的逻辑
        SingleOutputStreamOperator<Integer> minusOne = iterate.map(new MapFunction<Integer, Integer>() {
            @Override
            public Integer map(Integer integer) throws Exception {
                System.out.println("迭代体中进来的数据为 = "+integer);
                return integer - 1;
            }
        });

        //定义迭代条件 ，就是什么时候需要继续迭代计算
        SingleOutputStreamOperator<Integer> cn = minusOne.filter(new FilterFunction<Integer>() {
            @Override
            public boolean filter(Integer value) throws Exception {
                return value > 0;
            }
        });

        //对迭代流应用条件
        iterate.closeWith(cn);
        iterate.print();
        env.execute();

    }
}
